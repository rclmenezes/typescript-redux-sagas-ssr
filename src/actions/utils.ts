export type Error<T> = { [P in keyof T]?: string };

export type FormikSetErrors<T> = (errors: Error<T>) => void;

export interface FormikPayload<D> {
  data: D;
  setErrors: FormikSetErrors<D>;
}

export interface FormikTokenPayload<P> extends FormikPayload<P> {
  token: string;
}
