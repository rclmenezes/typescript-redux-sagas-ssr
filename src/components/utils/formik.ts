import { AxiosError } from "axios";

export type Error<T> = { [P in keyof T]?: string | undefined };

type FormikSetErrors<T> = (errors: T) => void;

export function setFormikErrorsFromError<T>(error: AxiosError, setErrors: FormikSetErrors<T>) {
  if (error.response && error.response.status === 400 && error.response.data) {
    setErrors(error.response.data);
  }
}
