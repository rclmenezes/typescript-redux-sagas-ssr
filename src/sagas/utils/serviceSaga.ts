import axios, { AxiosError, AxiosInstance } from "axios";
import { SagaIterator } from "redux-saga";
import { call, put, takeLatest } from "redux-saga/effects";
import { Action, AsyncActionCreators } from "typescript-fsa";

import { IS_CLIENT, PORT } from "../../settings";

let axiosInstance: AxiosInstance;
if (IS_CLIENT) {
  axiosInstance = axios.create();
} else {
  axiosInstance = axios.create({
    baseURL: `http://localhost:${PORT}`,
  });
}

export type Get = "GET";
export type Post = "POST";
export type MethodTypes = Get | Post;
export const GET: Get = "GET";
export const POST: Post = "POST";

export function createService<Data, Return>(url: string, method: MethodTypes) {
  return async (data: Data, headers: Record<string, string> | null) => {
    const res = await axiosInstance.request({
      data,
      headers,
      method,
      url,
    });
    return res.data as Return;
  };
}

export function createServiceSaga<P, S, E, D>(options: {
  actionCreator: AsyncActionCreators<P, S, E>;
  extractDataFromAction?: (action: Action<P>) => D;
  extractHeadersFromAction?: ((action: Action<P>) => Record<string, string>) | null;
  onError?: (action: Action<P>, e: AxiosError) => SagaIterator;
  onSuccess?: (action: Action<P>, result: S) => SagaIterator;
  method?: MethodTypes;
  url: string;
}) {
  const defaultOptions = {
    extractDataFromAction: (action: Action<P>) => (action.payload as unknown) as D,
    extractHeadersFromAction: null,
    method: GET,
  };
  const {
    actionCreator,
    extractDataFromAction,
    extractHeadersFromAction,
    onError,
    onSuccess,
    method,
    url,
  } = {
    ...defaultOptions,
    ...options,
  };

  const service = createService(url, method);

  const saga = function*(action: Action<P>) {
    const data = extractDataFromAction(action);
    const headers = extractHeadersFromAction ? extractHeadersFromAction(action) : null;
    try {
      const result: S = yield call(service, data, headers);
      yield put(
        actionCreator.done(
          {
            params: action.payload,
            result,
          },
          action.meta,
        ),
      );
      if (onSuccess) {
        yield* onSuccess(action, result);
      }
    } catch (err) {
      yield put(
        actionCreator.failed(
          {
            error: err,
            params: action.payload,
          },
          action.meta,
        ),
      );
      if (onError) {
        yield* onError(action, err);
      }
    }
  };
  return function*() {
    yield takeLatest<Action<P>>(actionCreator.started.type, saga);
  };
}
