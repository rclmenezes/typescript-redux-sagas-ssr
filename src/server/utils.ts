import { NextFunction, Request, Response } from "express";
// import sequelize from "../db/sequelize";

// export function wrapAsync<T>(fn: (req: Request, res: Response) => Promise<T>) {
//   return (req: Request, res: Response, next: NextFunction) => {
//     sequelize
//       .transaction(t => {
//         return fn(req, res);
//       })
//       .catch(next);
//   };
// }

export function wrapAsync<T>(fn: (req: Request, res: Response, next: NextFunction) => Promise<T>) {
  return (req: Request, res: Response, next: NextFunction) => {
    Promise.resolve(fn(req, res, next)).catch(next);
  };
}
