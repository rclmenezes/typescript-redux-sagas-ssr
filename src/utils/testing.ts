import { Href, UnregisterCallback } from "history";

import { PageRoutingProps } from "./routing";

export function createMockRouterProps<P>(params: P, search: string = "") {
  const location = {
    hash: "",
    key: "",
    pathname: "",
    search,
    state: {},
  };

  const props: PageRoutingProps<P> = {
    history: {
      action: "POP",
      block: t => {
        const temp: UnregisterCallback = () => {};
        return temp;
      },
      createHref: t => {
        const temp: Href = "";
        return temp;
      },
      go: num => {},
      goBack: () => {},
      goForward: () => {},
      length: 2,
      listen: t => {
        const temp: UnregisterCallback = () => {};
        return temp;
      },
      location,
      push: () => {},
      replace: () => {},
    },
    location,
    match: {
      isExact: true,
      params,
      path: "",
      url: "",
    },
    staticContext: {
      status: 200,
    },
  };
  return props;
}
