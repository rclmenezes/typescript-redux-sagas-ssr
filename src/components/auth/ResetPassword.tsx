import * as csstips from "csstips";
import { Field, Form, FormikBag, FormikProps, withFormik } from "formik";
import { decode } from "jsonwebtoken";
import { parse } from "query-string";
import * as React from "react";
import { Link } from "react-router-dom";
import { Button, Message } from "semantic-ui-react";
import { style } from "typestyle";
import * as Yup from "yup";

import { ResetPassword, ResetPasswordPayload } from "../../actions/auth";
import { UserState } from "../../reducers/user";
import { BASE_URL } from "../../settings";
import { PageRoutingProps } from "../../utils/routing";
import InputField from "../common/Input";
import BaseAuthPage from "./BaseAuthPage";

export interface ResetPasswordProps {
  resetPasswordSubmit: (payload: ResetPasswordPayload) => void;
  routing: PageRoutingProps<{}>;
  showConfirmation: boolean;
  user: UserState;
}

interface ResetPasswordFormProps extends ResetPasswordProps {
  token: string;
}

const InnerResetPasswordForm: React.SFC<
  FormikProps<ResetPassword> & ResetPasswordFormProps
> = () => (
  <Form>
    <Field
      name="password"
      component={InputField}
      placeholder="New Password"
      type="password"
      fluid
    />
    <Button type="submit" fluid primary>
      Submit
    </Button>
    <Message className={style(csstips.horizontallyCenterChildren)}>
      <Link to="/sign-up">Create an account</Link>
    </Message>
  </Form>
);

const ResetPasswordForm = withFormik<ResetPasswordFormProps, ResetPassword>({
  handleSubmit: (
    values: ResetPassword,
    formikBag: FormikBag<ResetPasswordFormProps, ResetPassword>,
  ) => {
    formikBag.props.resetPasswordSubmit({
      data: values,
      setErrors: formikBag.setErrors,
      token: formikBag.props.token,
    });
  },
  mapPropsToValues: () => {
    return {
      password: "",
    };
  },
  validationSchema: Yup.object().shape({
    password: Yup.string()
      .min(8, "Password must be at least 8 characters long")
      .required("Password is a required field"),
  }),
})(InnerResetPasswordForm);

type EXPIRED = "EXPIRED";
type INVALID = "INVALID";
type UNKNOWN = "UNKNOWN";
type VALID = "VALID";
type TokenStates = UNKNOWN | EXPIRED | VALID | INVALID;
const EXPIRED: EXPIRED = "EXPIRED";
const INVALID: INVALID = "INVALID";
const UNKNOWN: UNKNOWN = "UNKNOWN";
const VALID: VALID = "VALID";

class ResetPasswordPage extends React.Component<ResetPasswordProps> {
  componentDidMount = () => {
    const queryParams = parse(this.props.routing.location.search);
    let token = null;
    let tokenState: TokenStates = UNKNOWN;
    if (queryParams.token) {
      token = queryParams.token instanceof Array ? queryParams.token[0] : queryParams.token;
      const jwt = decode(token);
      if (jwt && typeof jwt === "object") {
        const isExpired = new Date().getTime() / 1000 > jwt.exp;
        tokenState = isExpired ? EXPIRED : VALID;
      } else {
        tokenState = INVALID;
      }
    }
    this.setState({ token: queryParams.token ? token : null, tokenState });
  };

  getToken() {
    const queryParams = parse(this.props.routing.location.search);
    let token = null;
    let tokenState: TokenStates = UNKNOWN;
    if (queryParams.token) {
      token = queryParams.token instanceof Array ? queryParams.token[0] : queryParams.token;
      const jwt = decode(token);
      if (jwt && typeof jwt === "object") {
        const isExpired = new Date().getTime() / 1000 > jwt.exp;
        tokenState = isExpired ? EXPIRED : VALID;
      } else {
        tokenState = INVALID;
      }
    }
    return { token: queryParams.token ? token : null, tokenState };
  }

  public render() {
    if (this.props.showConfirmation) {
      return (
        <BaseAuthPage title="Success!">
          <Message className={style(csstips.horizontallyCenterChildren)}>
            We've reset your password. Click <Link to="/">here to go home.</Link>
          </Message>
        </BaseAuthPage>
      );
    }

    const { token, tokenState } = this.getToken();
    if (tokenState !== VALID) {
      // If loading the page from here, make it 400
      const { staticContext } = this.props.routing;
      if (staticContext) {
        staticContext.status = 400;
      }
    }

    if (tokenState === INVALID || tokenState === UNKNOWN) {
      return (
        <BaseAuthPage title="Invalid URL">
          <Message className={style(csstips.horizontallyCenterChildren)}>
            Something is wrong with this URL. If you need help resetting your password, contact
            support at <a href="mailto:support@{BASE_URL}">support@{BASE_URL}</a>.
          </Message>
        </BaseAuthPage>
      );
    } else if (tokenState === EXPIRED) {
      return (
        <BaseAuthPage title="Expired URL">
          <Message className={style(csstips.horizontallyCenterChildren)}>
            This URL is expired. <Link to="/forgot-password">Please go here</Link> to submit another
            password request.
          </Message>
        </BaseAuthPage>
      );
    } else if (tokenState === VALID) {
      return (
        <BaseAuthPage title="Change Your Password">
          <ResetPasswordForm {...this.props} token={token!} />
        </BaseAuthPage>
      );
    } else {
      throw new Error("Invalid tokenState");
    }
  }
}

export default ResetPasswordPage;
