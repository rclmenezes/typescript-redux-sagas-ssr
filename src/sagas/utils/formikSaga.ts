import { AxiosError } from "axios";
import { call } from "redux-saga/effects";
import { Action, AsyncActionCreators } from "typescript-fsa";

import { FormikPayload } from "../../actions/utils";
import { createServiceSaga, POST } from "./serviceSaga";

export { FormikPayload };

export type FormikAction<P> = Action<FormikPayload<P>>;
export type FormikAsyncActionCreator<P, S, E> = AsyncActionCreators<FormikPayload<P>, S, E>;

export function* handleFormErrors<P>(action: FormikAction<P>, error: AxiosError) {
  if (error.response && error.response.status === 400 && error.response.data) {
    yield call(action.payload.setErrors, error.response.data);
  }
}

export function createFormikSubmitSaga<P, S, E>(
  actionCreator: FormikAsyncActionCreator<P, S, E>,
  url: string,
  extractHeadersFromAction: ((action: FormikAction<P>) => {}) | null = null,
) {
  return createServiceSaga({
    actionCreator,
    extractDataFromAction: action => action.payload.data,
    extractHeadersFromAction,
    method: POST,
    onError: handleFormErrors,
    url,
  });
}
