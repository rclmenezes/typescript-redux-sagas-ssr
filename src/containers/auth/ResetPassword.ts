import { connect } from "react-redux";
import { bindActionCreators, Dispatch } from "redux";
import actionCreators from "../../actions";
import ResetPassword from "../../components/auth/ResetPassword";
import { RootState } from "../../reducers";

const mapStateToProps = (state: RootState) => ({
  showConfirmation: state.resetPasswordSuccess,
  user: state.user,
});

const mapDispatchToProps = (dispatch: Dispatch) =>
  bindActionCreators(
    {
      resetPasswordSubmit: actionCreators.resetPassword.started,
    },
    dispatch,
  );

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(ResetPassword);
