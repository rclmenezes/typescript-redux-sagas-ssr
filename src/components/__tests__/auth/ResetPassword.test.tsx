import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import { sign } from "jsonwebtoken";
import React from "react";
import { MemoryRouter } from "react-router";

import { createMockRouterProps } from "../../../utils/testing";
import ResetPassword, { ResetPasswordProps } from "../../auth/ResetPassword";

describe("Test ResetPassword", () => {
  const MOCK_RESET_PASSWORD = {
    password: "12341234",
  };
  const FAKE_USER_ID = 1;
  const MOCK_VALID_TOKEN = sign({ id: FAKE_USER_ID }, "FAKE SECRET", { expiresIn: "1d" });

  const setup = (propOverrides: Partial<ResetPasswordProps> = {}, search = "") => {
    const resetPasswordSubmit = jest.fn();
    const routing = createMockRouterProps({}, search);

    const props = {
      resetPasswordSubmit,
      routing,
      showConfirmation: false,
      user: null,
      ...propOverrides,
    };

    const wrapper = mount(
      <MemoryRouter>
        <ResetPassword {...props} />
      </MemoryRouter>,
    );

    return {
      getFormProps: () => wrapper.find("InnerResetPasswordForm").props() as any,
      getPasswordField: () => wrapper.find('FieldInner[name="password"]'),
      resetPasswordSubmit,
      routing,
      setPasswordInput: (password: string) =>
        wrapper.find('input[name="password"]').simulate("change", {
          persist: () => {},
          target: {
            name: "password",
            value: password,
          },
        }),
      wrapper,
    };
  };

  test("Show confirmation", async () => {
    const { routing, wrapper } = setup({ showConfirmation: true });

    expect(routing.staticContext!.status).toBe(200);
    expect(wrapper.text()).toEqual(expect.stringContaining("Success"));
    expect(toJson(wrapper.find("BaseAuthPage") as any)).toMatchSnapshot();
  });

  test("Missing token", async () => {
    const { routing, wrapper } = setup();

    expect(routing.staticContext!.status).toBe(400);
    expect(wrapper.text()).toEqual(expect.stringContaining("Invalid"));
    expect(toJson(wrapper.find("BaseAuthPage") as any)).toMatchSnapshot();
  });

  test("Invalid token", async () => {
    const { routing, wrapper } = setup({}, "?token=invalidtoken");

    expect(routing.staticContext!.status).toBe(400);
    expect(wrapper.text()).toEqual(expect.stringContaining("Invalid"));
    expect(toJson(wrapper.find("BaseAuthPage") as any)).toMatchSnapshot();
  });

  test("Expired token", async () => {
    const token = sign({ id: FAKE_USER_ID }, "FAKE SECRET", { expiresIn: "-1d" });
    const { routing, wrapper } = setup({}, `?token=${token}`);

    expect(routing.staticContext!.status).toBe(400);
    expect(wrapper.text()).toEqual(expect.stringContaining("Expired"));
    expect(toJson(wrapper.find("BaseAuthPage") as any)).toMatchSnapshot();
  });

  test("Successful submit", async () => {
    const { getFormProps, resetPasswordSubmit, routing, setPasswordInput, wrapper } = setup(
      {},
      `?token=${MOCK_VALID_TOKEN}`,
    );

    expect(routing.staticContext!.status).toBe(200);
    expect(wrapper.text()).toEqual(expect.stringContaining("Change Your Password"));
    expect(toJson(wrapper.find("Form") as any)).toMatchSnapshot();

    setPasswordInput(MOCK_RESET_PASSWORD.password);

    let formikProps = getFormProps();
    await formikProps.submitForm();
    wrapper.update();
    formikProps = getFormProps();

    expect(formikProps.isSubmitting).toBe(true);
    expect(formikProps.isValid).toBe(true);
    expect(resetPasswordSubmit.mock.calls.length).toBe(1);
  });

  test("setErrors gets called", async () => {
    const {
      getFormProps,
      getPasswordField,
      resetPasswordSubmit,
      setPasswordInput,
      wrapper,
    } = setup({}, `?token=${MOCK_VALID_TOKEN}`);

    setPasswordInput(MOCK_RESET_PASSWORD.password);

    let formikProps = getFormProps();
    await formikProps.submitForm();
    wrapper.update();
    formikProps = getFormProps();

    expect(formikProps.isSubmitting).toBe(true);
    expect(formikProps.isValid).toBe(true);
    expect(resetPasswordSubmit.mock.calls.length).toBe(1);

    expect(resetPasswordSubmit.mock.calls[0][0].data).toEqual(MOCK_RESET_PASSWORD);

    const passwordError = "Badness";
    resetPasswordSubmit.mock.calls[0][0].setErrors({
      password: passwordError,
    });
    expect(getPasswordField().text()).toBe(passwordError);
  });

  test("Test submit without input", async () => {
    const { getFormProps, getPasswordField, wrapper } = setup({}, `?token=${MOCK_VALID_TOKEN}`);

    let formikProps = getFormProps();
    await formikProps.submitForm();
    wrapper.update();

    formikProps = getFormProps();
    expect(formikProps.errors.password).toBeTruthy();
    expect(getPasswordField().text()).toBe(formikProps.errors.password);
  });
});
