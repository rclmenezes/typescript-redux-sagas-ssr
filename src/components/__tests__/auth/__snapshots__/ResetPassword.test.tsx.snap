// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test ResetPassword Expired token 1`] = `
<BaseAuthPage
  title="Expired URL"
>
  <Container
    text={true}
  >
    <div
      className="ui text container"
    >
      <Header
        as="h1"
        textAlign="center"
      >
        <h1
          className="ui center aligned header"
        >
          Expired URL
        </h1>
      </Header>
      <Message
        className="fsuivmx"
      >
        <div
          className="ui message fsuivmx"
        >
          This URL is expired. 
          <Link
            replace={false}
            to="/forgot-password"
          >
            <a
              href="/forgot-password"
              onClick={[Function]}
            >
              Please go here
            </a>
          </Link>
           to submit another password request.
        </div>
      </Message>
    </div>
  </Container>
</BaseAuthPage>
`;

exports[`Test ResetPassword Invalid token 1`] = `
<BaseAuthPage
  title="Invalid URL"
>
  <Container
    text={true}
  >
    <div
      className="ui text container"
    >
      <Header
        as="h1"
        textAlign="center"
      >
        <h1
          className="ui center aligned header"
        >
          Invalid URL
        </h1>
      </Header>
      <Message
        className="fsuivmx"
      >
        <div
          className="ui message fsuivmx"
        >
          Something is wrong with this URL. If you need help resetting your password, contact support at 
          <a
            href="mailto:support@{BASE_URL}"
          >
            support@
            localhost:3000
          </a>
          .
        </div>
      </Message>
    </div>
  </Container>
</BaseAuthPage>
`;

exports[`Test ResetPassword Missing token 1`] = `
<BaseAuthPage
  title="Invalid URL"
>
  <Container
    text={true}
  >
    <div
      className="ui text container"
    >
      <Header
        as="h1"
        textAlign="center"
      >
        <h1
          className="ui center aligned header"
        >
          Invalid URL
        </h1>
      </Header>
      <Message
        className="fsuivmx"
      >
        <div
          className="ui message fsuivmx"
        >
          Something is wrong with this URL. If you need help resetting your password, contact support at 
          <a
            href="mailto:support@{BASE_URL}"
          >
            support@
            localhost:3000
          </a>
          .
        </div>
      </Message>
    </div>
  </Container>
</BaseAuthPage>
`;

exports[`Test ResetPassword Show confirmation 1`] = `
<BaseAuthPage
  title="Success!"
>
  <Container
    text={true}
  >
    <div
      className="ui text container"
    >
      <Header
        as="h1"
        textAlign="center"
      >
        <h1
          className="ui center aligned header"
        >
          Success!
        </h1>
      </Header>
      <Message
        className="fsuivmx"
      >
        <div
          className="ui message fsuivmx"
        >
          We've reset your password. Click 
          <Link
            replace={false}
            to="/"
          >
            <a
              href="/"
              onClick={[Function]}
            >
              here to go home.
            </a>
          </Link>
        </div>
      </Message>
    </div>
  </Container>
</BaseAuthPage>
`;

exports[`Test ResetPassword Successful submit 1`] = `
<Form>
  <Component
    formik={
      Object {
        "dirty": false,
        "errors": Object {},
        "handleBlur": [Function],
        "handleChange": [Function],
        "handleReset": [Function],
        "handleSubmit": [Function],
        "initialValues": Object {
          "password": "",
        },
        "isSubmitting": false,
        "isValid": false,
        "isValidating": false,
        "registerField": [Function],
        "resetForm": [Function],
        "setError": [Function],
        "setErrors": [Function],
        "setFieldError": [Function],
        "setFieldTouched": [Function],
        "setFieldValue": [Function],
        "setFormikState": [Function],
        "setStatus": [Function],
        "setSubmitting": [Function],
        "setTouched": [Function],
        "setValues": [Function],
        "submitCount": 0,
        "submitForm": [Function],
        "touched": Object {},
        "unregisterField": [Function],
        "validate": undefined,
        "validateField": [Function],
        "validateForm": [Function],
        "validateOnBlur": true,
        "validateOnChange": true,
        "validationSchema": [Function],
        "values": Object {
          "password": "",
        },
      }
    }
  >
    <form
      onReset={[Function]}
      onSubmit={[Function]}
    >
      <FormikConnect(FieldInner)
        component={[Function]}
        fluid={true}
        name="password"
        placeholder="New Password"
        type="password"
      >
        <FieldInner
          component={[Function]}
          fluid={true}
          formik={
            Object {
              "dirty": false,
              "errors": Object {},
              "handleBlur": [Function],
              "handleChange": [Function],
              "handleReset": [Function],
              "handleSubmit": [Function],
              "initialValues": Object {
                "password": "",
              },
              "isSubmitting": false,
              "isValid": false,
              "isValidating": false,
              "registerField": [Function],
              "resetForm": [Function],
              "setError": [Function],
              "setErrors": [Function],
              "setFieldError": [Function],
              "setFieldTouched": [Function],
              "setFieldValue": [Function],
              "setFormikState": [Function],
              "setStatus": [Function],
              "setSubmitting": [Function],
              "setTouched": [Function],
              "setValues": [Function],
              "submitCount": 0,
              "submitForm": [Function],
              "touched": Object {},
              "unregisterField": [Function],
              "validate": undefined,
              "validateField": [Function],
              "validateForm": [Function],
              "validateOnBlur": true,
              "validateOnChange": true,
              "validationSchema": [Function],
              "values": Object {
                "password": "",
              },
            }
          }
          name="password"
          placeholder="New Password"
          type="password"
        >
          <InputField
            field={
              Object {
                "name": "password",
                "onBlur": [Function],
                "onChange": [Function],
                "value": "",
              }
            }
            fluid={true}
            form={
              Object {
                "dirty": false,
                "errors": Object {},
                "handleBlur": [Function],
                "handleChange": [Function],
                "handleReset": [Function],
                "handleSubmit": [Function],
                "initialValues": Object {
                  "password": "",
                },
                "isSubmitting": false,
                "isValid": false,
                "isValidating": false,
                "registerField": [Function],
                "resetForm": [Function],
                "setError": [Function],
                "setErrors": [Function],
                "setFieldError": [Function],
                "setFieldTouched": [Function],
                "setFieldValue": [Function],
                "setFormikState": [Function],
                "setStatus": [Function],
                "setSubmitting": [Function],
                "setTouched": [Function],
                "setValues": [Function],
                "submitCount": 0,
                "submitForm": [Function],
                "touched": Object {},
                "unregisterField": [Function],
                "validateField": [Function],
                "validateForm": [Function],
                "validateOnBlur": true,
                "validateOnChange": true,
                "values": Object {
                  "password": "",
                },
              }
            }
            placeholder="New Password"
            type="password"
          >
            <div
              className="f1urtfz"
            >
              <Input
                fluid={true}
                name="password"
                onBlur={[Function]}
                onChange={[Function]}
                placeholder="New Password"
                type="password"
                value=""
              >
                <div
                  className="ui fluid input"
                >
                  <input
                    name="password"
                    onBlur={[Function]}
                    onChange={[Function]}
                    placeholder="New Password"
                    type="password"
                    value=""
                  />
                </div>
              </Input>
            </div>
          </InputField>
        </FieldInner>
      </FormikConnect(FieldInner)>
      <Button
        as="button"
        fluid={true}
        primary={true}
        type="submit"
      >
        <button
          className="ui fluid primary button"
          onClick={[Function]}
          type="submit"
        >
          Submit
        </button>
      </Button>
      <Message
        className="fsuivmx"
      >
        <div
          className="ui message fsuivmx"
        >
          <Link
            replace={false}
            to="/sign-up"
          >
            <a
              href="/sign-up"
              onClick={[Function]}
            >
              Create an account
            </a>
          </Link>
        </div>
      </Message>
    </form>
  </Component>
</Form>
`;
